src/System.Net.Security/tests/FunctionalTests/NegotiateStreamInvalidOperationTest.cs

```diff
@@ -0,0 +1,274 @@
+    public class NegotiateStreamInvalidOperationTest
+    {
+        private readonly byte[] _sampleMsg = Encoding.UTF8.GetBytes("Sample Test Message");
+
+        [Fact]
+        public async Task NegotiateStream_StreamContractTest_Success()
+        {
+            VirtualNetwork network = new VirtualNetwork();
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                Assert.False(client.CanSeek);
+                Assert.False(client.CanRead);
+                Assert.False(client.CanTimeout);
+                Assert.False(client.CanWrite);
+                Assert.False(server.CanSeek);
+                Assert.False(server.CanRead);
+                Assert.False(server.CanTimeout);
+                Assert.False(server.CanWrite);
+
+                Assert.Throws<InvalidOperationException>(() => client.ReadTimeout);
+                Assert.Throws<InvalidOperationException>(() => client.WriteTimeout);
+                Assert.Throws<NotImplementedException>(() => client.Length);
+                Assert.Throws<NotImplementedException>(() => client.Position);
+                Assert.Throws<NotSupportedException>(() => client.Seek(0, new SeekOrigin()));
+
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync();
+                auth[1] = server.AuthenticateAsServerAsync();
+
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                Assert.True(client.CanRead);
+                Assert.True(client.CanWrite);
+                Assert.True(server.CanRead);
+                Assert.True(server.CanWrite);
+            }
+        }
+
+        [Fact]
+        public async Task NegotiateStream_EndReadEndWriteInvalidParameter_Throws()
+        {
+            byte[] recvBuf = new byte[_sampleMsg.Length];
+            VirtualNetwork network = new VirtualNetwork();
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync(CredentialCache.DefaultNetworkCredentials, string.Empty);
+                auth[1] = server.AuthenticateAsServerAsync();
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(auth);
+
+                auth[0] = Task.Factory.FromAsync(client.BeginWrite, CustomEndWrite, _sampleMsg, 0, _sampleMsg.Length, client);
```

Comment:
It'd be easier to understand the test and what it's validating if the code in CustomEndWrite was put here as a lambda.

Reply:
Thanks for the suggestion, fixed.